marda_extractors_api/__init__.py

```diff
@@ -0,0 +1,155 @@
+import subprocess
+from enum import Enum
+from pathlib import Path
+
+import httpx
+
+__all__ = ("extract", "MardaExtractor")
+
+REGISTRY_BASE_URL = "https://marda-registry.fly.dev"
+
+
+class SupportedExecutionMethod(Enum):
+    CLI = "cli"
+    PYTHON = "python"
+
+
+def extract(
+    file_path: Path,
+    file_type: str,
+    output_file: Path | None = None,
+    preferred_mode: SupportedExecutionMethod = SupportedExecutionMethod.PYTHON,
+):
+    """Parse a file given its path and file type ID
+    in the MaRDA registry.
+    """
+
+    if not file_path.exists():
+        raise RuntimeError(f"File {file_path} does not exist")
+
+    response = httpx.get(f"{REGISTRY_BASE_URL}/filetypes/{file_type}")
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"Could not find file type {file_type!r} in the registry at {response.url!r}"
+        )
+    extractors = response.json()["registered_extractors"]
+    if not extractors:
+        raise RuntimeError(
+            f"No extractors found for file type {file_type!r} in the registry"
+        )
+    elif len(extractors) > 1:
+        print(
+            f"Discovered multiple extractors: {extractors}, using the first ({extractors[0]})"
+        )
+
+    extractor = extractors[0]
+    entry = httpx.get(f"{REGISTRY_BASE_URL}/extractors/{extractor}")
+    if response.status_code != 200:
+        raise RuntimeError(f"Could not find extractor {extractor!r} in the registry")
+
+    extractor = MardaExtractor(entry.json(), preferred_mode=preferred_mode)
+
+    return extractor.execute(file_type, file_path, output_file)
+
+
+class MardaExtractor:
+    """A plan for parsing a file."""
+
+    entry: dict
+    """The registry entry to use for parsing."""
+
+    def __init__(
+        self,
+        entry: dict,
+        install: bool = True,
+        preferred_mode: SupportedExecutionMethod = SupportedExecutionMethod.PYTHON,
+    ):
+        """Initialize the plan, optionally installing the specific parser package."""
+        self.entry = entry
+        self.preferred_mode = preferred_mode
+        if install:
+            self.install()
+
+    def install(self):
+        command = ["pip", "install", f"git+{self.entry['source_repository']}"]
+        subprocess.check_output(command)
```

Comment:
Adding my own comment here... some stuff missing to this "install" that we can think about:

- Add mechanism to install from PyPI, presumably with another field for machine readable installation instructions (another DSL, yay...), e.g., `pip:yadg` (could also be `pypi:yadg` or `pip[registry_url]:yadg` where the default is pypi), `conda[-c conda-forge]:yadg` (gross but might need to be more expressive with channels etc), `docker:ghcr.io/yadg`, `dockerfile:github.com/yadg/Dockerfile` etc.
- Then relatedly, add mechanism to check if already installed, for each of the above, so that the `install = False` flag can be removed (or augmented with `dont_install_if_missing` or something to allow users to force the issue).
- Add mechanism for isolating the install process. For Python, we could programmatically make a virtualenv for each extractor and then execute the final command inside that. I think we could probably do something similar inside a docker container (would be great to get a non-Python extractor added quickly so we don't overdesign the Python-specific stuff, could even write a noddy one myself)

Reply:
Should also think about minimum Python version for *this* package. Once we have isolated envs we can use whatever version but until then extractors have to support Py310 implicitly for now